// componentsjs/lib/rdf/Iris.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IRIS_OWL = exports.PREFIX_OWL = exports.IRIS_DOAP = exports.PREFIX_DOAP = exports.IRIS_XSD = exports.PREFIX_XSD = exports.IRIS_RDFS = exports.PREFIX_RDFS = exports.IRIS_RDF = exports.PREFIX_RDF = exports.PREFIX_OM = exports.IRIS_OO = exports.PREFIX_OO = void 0;
const definePrefix = (prefix) => (suffix) => `${prefix}${suffix}`;
exports.PREFIX_OO = definePrefix('https://linkedsoftwaredependencies.org/vocabularies/object-oriented#');
exports.IRIS_OO = {
    Module: (0, exports.PREFIX_OO)('Module'),
    Class: (0, exports.PREFIX_OO)('Class'),
    AbstractClass: (0, exports.PREFIX_OO)('AbstractClass'),
    ComponentInstance: (0, exports.PREFIX_OO)('ComponentInstance'),
    component: (0, exports.PREFIX_OO)('component'),
    componentPath: (0, exports.PREFIX_OO)('componentPath'),
    Override: (0, exports.PREFIX_OO)('Override'),
    overrideInstance: (0, exports.PREFIX_OO)('overrideInstance'),
    overrideParameters: (0, exports.PREFIX_OO)('overrideParameters'),
    parameter: (0, exports.PREFIX_OO)('parameter'),
};
exports.PREFIX_OM = definePrefix('https://linkedsoftwaredependencies.org/vocabularies/object-mapping#');
exports.PREFIX_RDF = definePrefix('http://www.w3.org/1999/02/22-rdf-syntax-ns#');
exports.IRIS_RDF = {
    subject: (0, exports.PREFIX_RDF)('subject'),
    object: (0, exports.PREFIX_RDF)('object'),
    type: (0, exports.PREFIX_RDF)('type'),
    JSON: (0, exports.PREFIX_RDF)('JSON'),
};
exports.PREFIX_RDFS = definePrefix('http://www.w3.org/2000/01/rdf-schema#');
exports.IRIS_RDFS = {
    seeAlso: (0, exports.PREFIX_RDFS)('seeAlso'),
};
exports.PREFIX_XSD = definePrefix('http://www.w3.org/2001/XMLSchema#');
exports.IRIS_XSD = {
    string: (0, exports.PREFIX_XSD)('string'),
    boolean: (0, exports.PREFIX_XSD)('boolean'),
    integer: (0, exports.PREFIX_XSD)('integer'),
    number: (0, exports.PREFIX_XSD)('number'),
    int: (0, exports.PREFIX_XSD)('int'),
    byte: (0, exports.PREFIX_XSD)('byte'),
    long: (0, exports.PREFIX_XSD)('long'),
    float: (0, exports.PREFIX_XSD)('float'),
    decimal: (0, exports.PREFIX_XSD)('decimal'),
    double: (0, exports.PREFIX_XSD)('double'),
};
exports.PREFIX_DOAP = definePrefix('http://usefulinc.com/ns/doap#');
exports.IRIS_DOAP = {
    name: (0, exports.PREFIX_DOAP)('name'),
};
exports.PREFIX_OWL = definePrefix('http://www.w3.org/2002/07/owl#');
exports.IRIS_OWL = {
    Restriction: (0, exports.PREFIX_OWL)('Restriction'),
};
