// @comunica/actor-rdf-join-inner-none/lib/ActorRdfJoinNone.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActorRdfJoinNone = void 0;
const bindings_factory_1 = require("@comunica/bindings-factory");
const bus_rdf_join_1 = require("@comunica/bus-rdf-join");
const asynciterator_1 = require("asynciterator");
const BF = new bindings_factory_1.BindingsFactory();
/**
 * A comunica None RDF Join Actor.
 */
class ActorRdfJoinNone extends bus_rdf_join_1.ActorRdfJoin {
    constructor(args) {
        super(args, {
            logicalType: 'inner',
            physicalName: 'none',
            limitEntries: 0,
        });
    }
    async test(action) {
        // Allow joining of one or zero streams
        if (action.entries.length > 0) {
            throw new Error(`Actor ${this.name} can only join zero entries`);
        }
        return await this.getJoinCoefficients();
    }
    async getOutput(action) {
        return {
            result: {
                bindingsStream: new asynciterator_1.ArrayIterator([BF.bindings()], { autoStart: false }),
                metadata: () => Promise.resolve({
                    cardinality: { type: 'exact', value: 1 },
                    canContainUndefs: false,
                    variables: [],
                }),
                type: 'bindings',
            },
        };
    }
    async getJoinCoefficients() {
        return {
            iterations: 0,
            persistedItems: 0,
            blockingItems: 0,
            requestTime: 0,
        };
    }
}
exports.ActorRdfJoinNone = ActorRdfJoinNone;
