// @comunica/actor-query-operation-nop/lib/ActorQueryOperationNop.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActorQueryOperationNop = void 0;
const bindings_factory_1 = require("@comunica/bindings-factory");
const bus_query_operation_1 = require("@comunica/bus-query-operation");
const asynciterator_1 = require("asynciterator");
const BF = new bindings_factory_1.BindingsFactory();
/**
 * A [Query Operation](https://github.com/comunica/comunica/tree/master/packages/bus-query-operation)
 * actor that handles SPARQL nop operations.
 */
class ActorQueryOperationNop extends bus_query_operation_1.ActorQueryOperationTypedMediated {
    constructor(args) {
        super(args, 'nop');
    }
    async testOperation(operation, context) {
        return true;
    }
    async runOperation(operation, context) {
        return {
            bindingsStream: new asynciterator_1.SingletonIterator(BF.bindings()),
            metadata: () => Promise.resolve({
                cardinality: { type: 'exact', value: 1 },
                canContainUndefs: false,
                variables: [],
            }),
            type: 'bindings',
        };
    }
}
exports.ActorQueryOperationNop = ActorQueryOperationNop;
