// @inrupt/solid-client-authn-node/dist/login/oidc/oidcHandlers/OidcHandlerCanHandleTests.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const IOidcOptions_1 = require("../__mocks__/IOidcOptions");
const canHandleTests = {
    legacyImplicitFlowOidcHandler: [
        {
            message: "should accept a configuration with many grant types including implicit",
            shouldPass: true,
            oidcOptions: {
                ...IOidcOptions_1.standardOidcOptions,
                issuerConfiguration: {
                    ...IOidcOptions_1.standardOidcOptions.issuerConfiguration,
                    grantTypesSupported: ["authorization_code", "implicit", "device"],
                },
            },
        },
        {
            message: "should accept a configuration with only the implicit grant type",
            shouldPass: true,
            oidcOptions: {
                ...IOidcOptions_1.standardOidcOptions,
                issuerConfiguration: {
                    ...IOidcOptions_1.standardOidcOptions.issuerConfiguration,
                    grantTypesSupported: ["implicit"],
                },
            },
        },
        {
            message: "shouldn't accept a configuration that has many grant types not including implicit",
            shouldPass: false,
            oidcOptions: {
                ...IOidcOptions_1.standardOidcOptions,
                issuerConfiguration: {
                    ...IOidcOptions_1.standardOidcOptions.issuerConfiguration,
                    grantTypesSupported: ["authorization_code", "device"],
                },
            },
        },
        {
            message: "shouldn't accept a configuration that does not include grant types",
            shouldPass: false,
            oidcOptions: IOidcOptions_1.standardOidcOptions,
        },
    ],
    authorizationCodeWithPkceOidcHandler: [
        {
            message: "should accept a configuration with many grant types including authorization code",
            shouldPass: true,
            oidcOptions: {
                ...IOidcOptions_1.standardOidcOptions,
                issuerConfiguration: {
                    ...IOidcOptions_1.standardOidcOptions.issuerConfiguration,
                    grantTypesSupported: ["authorization_code", "implicit", "device"],
                },
            },
        },
        {
            message: "should accept a configuration with only the authorization code grant type",
            shouldPass: true,
            oidcOptions: {
                ...IOidcOptions_1.standardOidcOptions,
                issuerConfiguration: {
                    ...IOidcOptions_1.standardOidcOptions.issuerConfiguration,
                    grantTypesSupported: ["authorization_code"],
                },
            },
        },
        {
            message: "shouldn't accept a configuration that has many grant types not including authorization code",
            shouldPass: false,
            oidcOptions: {
                ...IOidcOptions_1.standardOidcOptions,
                issuerConfiguration: {
                    ...IOidcOptions_1.standardOidcOptions.issuerConfiguration,
                    grantTypesSupported: ["implicit", "device"],
                },
            },
        },
        {
            message: "shouldn't accept a configuration that does not include grant types",
            shouldPass: false,
            oidcOptions: IOidcOptions_1.standardOidcOptions,
        },
    ],
};
exports.default = canHandleTests;
