// @comunica/actor-rdf-resolve-quad-pattern-hypermedia/lib/ActorRdfResolveQuadPatternHypermedia.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActorRdfResolveQuadPatternHypermedia = void 0;
const bus_rdf_resolve_quad_pattern_1 = require("@comunica/bus-rdf-resolve-quad-pattern");
const LRUCache = require("lru-cache");
const MediatedQuadSource_1 = require("./MediatedQuadSource");
/**
 * A comunica Hypermedia RDF Resolve Quad Pattern Actor.
 */
class ActorRdfResolveQuadPatternHypermedia extends bus_rdf_resolve_quad_pattern_1.ActorRdfResolveQuadPatternSource {
    constructor(args) {
        super(args);
        this.cache = this.cacheSize ? new LRUCache({ max: this.cacheSize }) : undefined;
        const cache = this.cache;
        if (cache) {
            this.httpInvalidator.addInvalidateListener(({ url }) => url ? cache.delete(url) : cache.clear());
        }
    }
    async test(action) {
        const sources = (0, bus_rdf_resolve_quad_pattern_1.hasContextSingleSource)(action.context);
        if (!sources) {
            throw new Error(`Actor ${this.name} can only resolve quad pattern queries against a single source.`);
        }
        return true;
    }
    getSource(context, operation) {
        const contextSource = (0, bus_rdf_resolve_quad_pattern_1.getContextSource)(context);
        const url = (0, bus_rdf_resolve_quad_pattern_1.getContextSourceUrl)(contextSource);
        let source;
        // Try to read from cache
        if (this.cache && this.cache.has(url)) {
            source = this.cache.get(url);
        }
        else {
            // If not in cache, create a new source
            source = new MediatedQuadSource_1.MediatedQuadSource(this.cacheSize, url, (0, bus_rdf_resolve_quad_pattern_1.getDataSourceType)(contextSource), this.maxIterators, this.aggregateStore, {
                mediatorMetadata: this.mediatorMetadata,
                mediatorMetadataExtract: this.mediatorMetadataExtract,
                mediatorDereferenceRdf: this.mediatorDereferenceRdf,
                mediatorRdfResolveHypermedia: this.mediatorRdfResolveHypermedia,
                mediatorRdfResolveHypermediaLinks: this.mediatorRdfResolveHypermediaLinks,
                mediatorRdfResolveHypermediaLinksQueue: this.mediatorRdfResolveHypermediaLinksQueue,
            });
            // Set in cache
            if (this.cache) {
                this.cache.set(url, source);
            }
        }
        return Promise.resolve(source);
    }
}
exports.ActorRdfResolveQuadPatternHypermedia = ActorRdfResolveQuadPatternHypermedia;
