// @comunica/logger-pretty/lib/LoggerPretty.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoggerPretty = void 0;
const types_1 = require("@comunica/types");
const objectInspect = require("object-inspect");
/**
 * A logger that pretty-prints everything.
 */
class LoggerPretty extends types_1.Logger {
    constructor(args) {
        super();
        this.level = args.level;
        this.levelOrdinal = types_1.Logger.getLevelOrdinal(this.level);
        this.actors = args.actors;
    }
    debug(message, data) {
        this.log('debug', message, data);
    }
    error(message, data) {
        this.log('error', message, data);
    }
    fatal(message, data) {
        this.log('fatal', message, data);
    }
    info(message, data) {
        this.log('info', message, data);
    }
    trace(message, data) {
        this.log('trace', message, data);
    }
    warn(message, data) {
        this.log('warn', message, data);
    }
    log(level, message, data) {
        if (types_1.Logger.getLevelOrdinal(level) >= this.levelOrdinal &&
            (!data || !('actor' in data) || !this.actors || this.actors[data.actor])) {
            process.stderr.write(`[${new Date().toISOString()}]  ${level.toUpperCase()}: ${message} ${objectInspect(data)}\n`);
        }
    }
}
exports.LoggerPretty = LoggerPretty;
