// @comunica/query-sparql-solid/lib/CliArgsHandlerSolidAuth.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CliArgsHandlerSolidAuth = void 0;
const { interactiveLogin } = require('solid-node-interactive-auth');
/**
 * Adds and handles CLI options for Solid authentication.
 */
class CliArgsHandlerSolidAuth {
    populateYargs(argumentsBuilder) {
        return argumentsBuilder
            .options({
            identityProvider: {
                alias: 'idp',
                type: 'string',
                describe: 'Solid identity provider to authenticate with (set to \'void\' to disable auth)',
                default: 'https://solidcommunity.net/',
                group: 'Recommended options:',
            },
        });
    }
    async handleArgs(args, context) {
        if (args.identityProvider !== 'void') {
            this.session = await interactiveLogin({
                oidcIssuer: args.identityProvider,
            });
            context['@comunica/actor-http-inrupt-solid-client-authn:session'] = this.session;
        }
    }
}
exports.CliArgsHandlerSolidAuth = CliArgsHandlerSolidAuth;
